cmd/cstorvolumeclaim/controller.go

```diff
@@ -385,3 +398,199 @@ func BaseLabels(cvc *apis.CStorVolumeClaim) map[string]string {
+		return false
+	}
+
+	requestCVCSize := cvc.Spec.Capacity[v1.ResourceStorage]
+	actualCVCSize := cvc.Status.Capacity[v1.ResourceStorage]
+	if requestCVCSize.Cmp(actualCVCSize) > 0 {
+		return true
+	}
+
+	actualCVSize := cv.Spec.Capacity
+	return requestCVCSize.Cmp(actualCVSize) > 0
+}
+
+// resizeCVC will:
+// 1. Mark cvc as resizing.
+// 2. Resize the cstorvolume object.
+// 3. Mark cvc as resizing finished
+func (c *CVCController) resizeCVC(cvc *apis.CStorVolumeClaim) error {
+	var updatedCVC *apis.CStorVolumeClaim
+	var err error
+	cv, err := c.clientset.OpenebsV1alpha1().CStorVolumes(cvc.Namespace).
+		Get(cvc.Name, metav1.GetOptions{})
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("falied to get cv %s: %v", cvc.Name, err))
+		return err
+	}
+
+	requestCVCSize := cvc.Spec.Capacity[v1.ResourceStorage]
+	if requestCVCSize.Cmp(cv.Spec.Capacity) > 0 {
+		if updatedCVC, err = c.markCVCResizeInProgress(cvc); err != nil {
+			glog.Errorf("failed to mark cvc %q as resizing: %v", cvc.Name, err)
+			return err
+		}
+		cvc = updatedCVC
+		// Record an event to indicate that cvc-controller is resizing this volume.
+		c.recorder.Event(cvc, corev1.EventTypeNormal, string(apis.CStorVolumeClaimResizing),
+			fmt.Sprintf("CVCController is resizing volume %s", cvc.Name))
+
+		err = c.resizeVolume(cvc, cv)
+		if err != nil {
+			// Record an event to indicate that resize operation is failed.
+			c.recorder.Eventf(cvc, corev1.EventTypeWarning, string(apis.CStorVolumeClaimResizeFailed), err.Error())
+		}
+		return err
+	}
+	if requestCVCSize.Cmp(cv.Status.Capacity) == 0 {
+		// Resize volume succeeded mark it as resizing finished.
+		return c.markCVCResizeFinished(cvc)
+	}
+	return nil
+}
+
+func (c *CVCController) markCVCResizeInProgress(cvc *apis.CStorVolumeClaim) (*apis.CStorVolumeClaim, error) {
+	// Mark CVC as Resize Started
+	progressCondition := apis.CStorVolumeClaimCondition{
+		Type:               apis.CStorVolumeClaimResizing,
+		LastTransitionTime: metav1.Now(),
+	}
+	newCVC := cvc.DeepCopy()
+	newCVC.Status.Conditions = MergeResizeConditionsOfCVC(newCVC.Status.Conditions,
+		[]apis.CStorVolumeClaimCondition{progressCondition})
+	return c.PatchCVCStatus(cvc, newCVC)
+}
+
+type resizeProcessStatus struct {
+	condition apis.CStorVolumeClaimCondition
+	processed bool
+}
+
+// MergeResizeConditionsOfCVC updates cvc with requested resize conditions
+// leaving other conditions untouched.
+func MergeResizeConditionsOfCVC(oldConditions, resizeConditions []apis.CStorVolumeClaimCondition) []apis.CStorVolumeClaimCondition {
+
+	resizeConditionMap := map[apis.CStorVolumeClaimConditionType]*resizeProcessStatus{}
+
+	for _, condition := range resizeConditions {
+		resizeConditionMap[condition.Type] = &resizeProcessStatus{condition, false}
+	}
+
+	newConditions := []apis.CStorVolumeClaimCondition{}
+	for _, condition := range oldConditions {
+		// If Condition is of not resize type, we keep it.
+		if _, ok := knownResizeConditions[condition.Type]; !ok {
```

Comment:
I think this kind of check will cause problem in future when we need to support multiple condition other than resize then when we trying to remove one then it will remove all other known condition from `cvc status`.

Reply:
this is only for resize, as we don't have any other conditions currently. later we can enhance it to merge different conditions as well..